File: Backend/src/firebase_service.py
"""
Firebase Firestore Service for DDI Predictor Backend
Handles all Firebase operations including user management and data storage
"""

import firebase_admin
from firebase_admin import credentials, firestore, auth
from datetime import datetime
from typing import Dict, List, Optional
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()


class FirebaseService:
    """Firebase Firestore service for DDI Predictor"""
    
    def __init__(self, credentials_path: str = None):
        """
        Initialize Firebase Admin SDK
        
        Args:
            credentials_path: Path to Firebase service account JSON file
        """
        if credentials_path is None:
            credentials_path = os.getenv('FIREBASE_CREDENTIALS_PATH', '../firebase-credentials.json')
        
        # Initialize Firebase Admin SDK (only once)
        if not firebase_admin._apps:
            try:
                cred = credentials.Certificate(credentials_path)
                firebase_admin.initialize_app(cred)
                print("✅ Firebase Admin SDK initialized successfully")
            except Exception as e:
                print(f"❌ Error initializing Firebase: {e}")
                raise
        
        # Get Firestore client
        self.db = firestore.client()
        print("✅ Firestore client connected")
    
    # ==================== User Management ====================
    
    def create_user(self, email: str, password: str, display_name: str = None) -> Dict:
        """
        Create a new user in Firebase Authentication and Firestore
        
        Args:
            email: User email
            password: User password
            display_name: Optional display name
            
        Returns:
            User data dictionary
        """
        try:
            # Create user in Firebase Auth
            user = auth.create_user(
                email=email,
                password=password,
                display_name=display_name
            )
            
            # Create user document in Firestore
            user_data = {
                'email': email,
                'displayName': display_name or email.split('@')[0],
                'createdAt': firestore.SERVER_TIMESTAMP,
                'totalChecks': 0,
                'preferences': {
                    'notifications': True,
                    'theme': 'light'
                }
            }
            
            self.db.collection('users').document(user.uid).set(user_data)
            
            print(f"✅ User created: {email}")
            return {'uid': user.uid, **user_data}
            
        except Exception as e:
            print(f"❌ Error creating user: {e}")
            raise
    
    def get_user(self, user_id: str) -> Optional[Dict]:
        """
        Get user data from Firestore
        
        Args:
            user_id: Firebase user ID
            
        Returns:
            User data dictionary or None
        """
        try:
            doc = self.db.collection('users').document(user_id).get()
            if doc.exists:
                return {'uid': user_id, **doc.to_dict()}
            return None
        except Exception as e:
            print(f"❌ Error getting user: {e}")
            return None
    
    def update_user(self, user_id: str, data: Dict) -> bool:
        """
        Update user data in Firestore
        
        Args:
            user_id: Firebase user ID
            data: Dictionary of fields to update
            
        Returns:
            True if successful
        """
        try:
            self.db.collection('users').document(user_id).update(data)
            print(f"✅ User updated: {user_id}")
            return True
        except Exception as e:
            print(f"❌ Error updating user: {e}")
            return False
    
    # ==================== Interaction Management ====================
    
    def save_interaction(self, interaction_data: Dict) -> str:
        """
        Save drug interaction result to Firestore
        
        Args:
            interaction_data: Dictionary containing interaction details
            
        Returns:
            Document ID of saved interaction
        """
        try:
            # Add timestamp if not present
            if 'timestamp' not in interaction_data:
                interaction_data['timestamp'] = firestore.SERVER_TIMESTAMP
            
            # Save to Firestore
            doc_ref = self.db.collection('interactions').add(interaction_data)
            doc_id = doc_ref[1].id
            
            # Update user's total checks count
            if 'userId' in interaction_data:
                user_ref = self.db.collection('users').document(interaction_data['userId'])
                user_ref.update({
                    'totalChecks': firestore.Increment(1)
                })
            
            print(f"✅ Interaction saved: {doc_id}")
            return doc_id
            
        except Exception as e:
            print(f"❌ Error saving interaction: {e}")
            raise
    
    def get_user_interactions(self, user_id: str, limit: int = 50) -> List[Dict]:
        """
        Get user's interaction history
        
        Args:
            user_id: Firebase user ID
            limit: Maximum number of interactions to return
            
        Returns:
            List of interaction dictionaries
        """
        try:
            docs = (self.db.collection('interactions')
                   .where('userId', '==', user_id)
                   .order_by('timestamp', direction=firestore.Query.DESCENDING)
                   .limit(limit)
                   .stream())
            
            interactions = []
            for doc in docs:
                data = doc.to_dict()
                data['id'] = doc.id
                interactions.append(data)
            
            return interactions
            
        except Exception as e:
            print(f"❌ Error getting interactions: {e}")
            return []
    
    def delete_interaction(self, interaction_id: str, user_id: str) -> bool:
        """
        Delete an interaction (with user verification)
        
        Args:
            interaction_id: Interaction document ID
            user_id: User ID (for verification)
            
        Returns:
            True if successful
        """
        try:
            doc_ref = self.db.collection('interactions').document(interaction_id)
            doc = doc_ref.get()
            
            if not doc.exists:
                return False
            
            # Verify ownership
            if doc.to_dict().get('userId') != user_id:
                print("❌ User doesn't own this interaction")
                return False
            
            doc_ref.delete()
            print(f"✅ Interaction deleted: {interaction_id}")
            return True
            
        except Exception as e:
            print(f"❌ Error deleting interaction: {e}")
            return False
    
    def clear_user_history(self, user_id: str) -> int:
        """
        Clear all interactions for a user
        
        Args:
            user_id: Firebase user ID
            
        Returns:
            Number of interactions deleted
        """
        try:
            docs = (self.db.collection('interactions')
                   .where('userId', '==', user_id)
                   .stream())
            
            count = 0
            batch = self.db.batch()
            
            for doc in docs:
                batch.delete(doc.reference)
                count += 1
            
            batch.commit()
            
            # Reset user's total checks
            self.db.collection('users').document(user_id).update({
                'totalChecks': 0
            })
            
            print(f"✅ Cleared {count} interactions for user {user_id}")
            return count
            
        except Exception as e:
            print(f"❌ Error clearing history: {e}")
            return 0
    
    # ==================== Drug Management ====================
    
    def add_drug(self, drug_data: Dict) -> str:
        """
        Add a drug to Firestore
        
        Args:
            drug_data: Dictionary containing drug information
            
        Returns:
            Document ID
        """
        try:
            doc_ref = self.db.collection('drugs').add(drug_data)
            doc_id = doc_ref[1].id
            print(f"✅ Drug added: {drug_data.get('name')}")
            return doc_id
        except Exception as e:
            print(f"❌ Error adding drug: {e}")
            raise
    
    def get_drug(self, drug_name: str) -> Optional[Dict]:
        """
        Get drug information by name
        
        Args:
            drug_name: Name of the drug
            
        Returns:
            Drug data dictionary or None
        """
        try:
            docs = (self.db.collection('drugs')
                   .where('name', '==', drug_name)
                   .limit(1)
                   .stream())
            
            for doc in docs:
                data = doc.to_dict()
                data['id'] = doc.id
                return data
            
            return None
            
        except Exception as e:
            print(f"❌ Error getting drug: {e}")
            return None
    
    def search_drugs(self, query: str, limit: int = 20) -> List[Dict]:
        """
        Search for drugs by name (prefix search)
        
        Args:
            query: Search query
            limit: Maximum results
            
        Returns:
            List of matching drugs
        """
        try:
            # Firestore doesn't support full-text search natively
            # This is a simple prefix search
            query_upper = query.upper()
            
            docs = (self.db.collection('drugs')
                   .where('name', '>=', query)
                   .where('name', '<=', query + '\uf8ff')
                   .limit(limit)
                   .stream())
            
            drugs = []
            for doc in docs:
                data = doc.to_dict()
                data['id'] = doc.id
                drugs.append(data)
            
            return drugs
            
        except Exception as e:
            print(f"❌ Error searching drugs: {e}")
            return []
    
    # ==================== Analytics ====================
    
    def save_analytics(self, date: str, analytics_data: Dict) -> bool:
        """
        Save daily analytics data
        
        Args:
            date: Date string (YYYY-MM-DD)
            analytics_data: Analytics dictionary
            
        Returns:
            True if successful
        """
        try:
            self.db.collection('analytics').document(date).set(
                analytics_data,
                merge=True
            )
            print(f"✅ Analytics saved for {date}")
            return True
        except Exception as e:
            print(f"❌ Error saving analytics: {e}")
            return False
    
    def get_analytics(self, start_date: str, end_date: str) -> List[Dict]:
        """
        Get analytics data for date range
        
        Args:
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)
            
        Returns:
            List of analytics dictionaries
        """
        try:
            docs = (self.db.collection('analytics')
                   .where(firestore.FieldPath.document_id(), '>=', start_date)
                   .where(firestore.FieldPath.document_id(), '<=', end_date)
                   .stream())
            
            analytics = []
            for doc in docs:
                data = doc.to_dict()
                data['date'] = doc.id
                analytics.append(data)
            
            return analytics
            
        except Exception as e:
            print(f"❌ Error getting analytics: {e}")
            return []
    
    # ==================== Utility Methods ====================
    
    def test_connection(self) -> bool:
        """
        Test Firestore connection
        
        Returns:
            True if connection is working
        """
        try:
            # Try to read from a collection
            test_ref = self.db.collection('_test').document('connection')
            test_ref.set({'timestamp': firestore.SERVER_TIMESTAMP})
            
            doc = test_ref.get()
            if doc.exists:
                test_ref.delete()
                print("✅ Firestore connection test successful")
                return True
            
            return False
            
        except Exception as e:
            print(f"❌ Firestore connection test failed: {e}")
            return False


# Example usage
if __name__ == "__main__":
    print("="*60)
    print("Firebase Service Test")
    print("="*60)
    
    try:
        # Initialize service
        firebase = FirebaseService()
        
        # Test connection
        if firebase.test_connection():
            print("\n✅ Firebase service is working correctly!")
        else:
            print("\n❌ Firebase service test failed")
    
    except Exception as e:
        print(f"\n❌ Error: {e}")
        print("\nMake sure:")
        print("1. firebase-credentials.json exists")
        print("2. Firebase project is set up")
        print("3. Firestore is enabled")

File: flutter/lib/services/firebase_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import '../models/interaction_result.dart';

class FirebaseService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;

  // ==================== Authentication ====================

  /// Get current user
  User? get currentUser => _auth.currentUser;

  /// Check if user is signed in
  bool get isSignedIn => _auth.currentUser != null;

  /// Sign in with email and password
  Future<UserCredential> signIn(String email, String password) async {
    try {
      return await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
    } catch (e) {
      throw Exception('Sign in failed: $e');
    }
  }

  /// Sign up with email and password
  Future<UserCredential> signUp(String email, String password) async {
    try {
      final credential = await _auth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );

      // Create user document in Firestore
      await _firestore.collection('users').doc(credential.user!.uid).set({
        'email': email,
        'displayName': email.split('@')[0],
        'createdAt': FieldValue.serverTimestamp(),
        'totalChecks': 0,
        'preferences': {
          'notifications': true,
          'theme': 'light',
        },
      });

      return credential;
    } catch (e) {
      throw Exception('Sign up failed: $e');
    }
  }

  /// Sign out
  Future<void> signOut() async {
    await _auth.signOut();
  }

  /// Sign in anonymously
  Future<UserCredential> signInAnonymously() async {
    return await _auth.signInAnonymously();
  }

  // ==================== User Data ====================

  /// Get user data
  Future<Map<String, dynamic>?> getUserData(String userId) async {
    try {
      final doc = await _firestore.collection('users').doc(userId).get();
      if (doc.exists) {
        return doc.data();
      }
      return null;
    } catch (e) {
      print('Error getting user data: $e');
      return null;
    }
  }

  /// Update user data
  Future<void> updateUserData(String userId, Map<String, dynamic> data) async {
    await _firestore.collection('users').doc(userId).update(data);
  }

  /// Increment total checks
  Future<void> incrementTotalChecks(String userId) async {
    await _firestore.collection('users').doc(userId).update({
      'totalChecks': FieldValue.increment(1),
    });
  }

  // ==================== Interaction History ====================

  /// Save interaction to Firestore
  Future<String> saveInteraction(InteractionResult result) async {
    try {
      final userId = currentUser?.uid;
      if (userId == null) {
        throw Exception('User not signed in');
      }

      final docRef = await _firestore.collection('interactions').add({
        'userId': userId,
        'drugPair': result.drugPair,
        'severity': result.severity,
        'riskScore': result.riskScore,
        'description': result.description,
        'mechanism': result.mechanism,
        'recommendations': result.recommendations,
        'sources': result.sources,
        'timestamp': FieldValue.serverTimestamp(),
      });

      // Increment user's total checks
      await incrementTotalChecks(userId);

      return docRef.id;
    } catch (e) {
      throw Exception('Failed to save interaction: $e');
    }
  }

  /// Get user's interaction history as stream
  Stream<List<InteractionResult>> getInteractionHistoryStream({int limit = 50}) {
    final userId = currentUser?.uid;
    if (userId == null) {
      return Stream.value([]);
    }

    return _firestore
        .collection('interactions')
        .where('userId', isEqualTo: userId)
        .orderBy('timestamp', descending: true)
        .limit(limit)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) {
        final data = doc.data();
        return InteractionResult(
          success: true,
          drugPair: data['drugPair'] ?? '',
          interactionExists: data['severity'] != 'None',
          severity: data['severity'] ?? 'Unknown',
          riskScore: (data['riskScore'] ?? 0).toDouble(),
          description: data['description'] ?? '',
          mechanism: data['mechanism'] ?? '',
          recommendations: List<String>.from(data['recommendations'] ?? []),
          sources: List<String>.from(data['sources'] ?? []),
          timestamp: (data['timestamp'] as Timestamp?)?.toDate().toString() ??
              DateTime.now().toString(),
        );
      }).toList();
    });
  }

  /// Get user's interaction history (one-time fetch)
  Future<List<InteractionResult>> getInteractionHistory({int limit = 50}) async {
    final userId = currentUser?.uid;
    if (userId == null) {
      return [];
    }

    try {
      final snapshot = await _firestore
          .collection('interactions')
          .where('userId', isEqualTo: userId)
          .orderBy('timestamp', descending: true)
          .limit(limit)
          .get();

      return snapshot.docs.map((doc) {
        final data = doc.data();
        return InteractionResult(
          success: true,
          drugPair: data['drugPair'] ?? '',
          interactionExists: data['severity'] != 'None',
          severity: data['severity'] ?? 'Unknown',
          riskScore: (data['riskScore'] ?? 0).toDouble(),
          description: data['description'] ?? '',
          mechanism: data['mechanism'] ?? '',
          recommendations: List<String>.from(data['recommendations'] ?? []),
          sources: List<String>.from(data['sources'] ?? []),
          timestamp: (data['timestamp'] as Timestamp?)?.toDate().toString() ??
              DateTime.now().toString(),
        );
      }).toList();
    } catch (e) {
      print('Error getting history: $e');
      return [];
    }
  }

  /// Delete a specific interaction
  Future<void> deleteInteraction(String interactionId) async {
    await _firestore.collection('interactions').doc(interactionId).delete();
  }

  /// Clear all user's interactions
  Future<int> clearHistory() async {
    final userId = currentUser?.uid;
    if (userId == null) {
      return 0;
    }

    try {
      final snapshot = await _firestore
          .collection('interactions')
          .where('userId', isEqualTo: userId)
          .get();

      final batch = _firestore.batch();
      for (var doc in snapshot.docs) {
        batch.delete(doc.reference);
      }
      await batch.commit();

      // Reset total checks
      await _firestore.collection('users').doc(userId).update({
        'totalChecks': 0,
      });

      return snapshot.docs.length;
    } catch (e) {
      print('Error clearing history: $e');
      return 0;
    }
  }

  // ==================== Statistics ====================

  /// Get user statistics
  Future<Map<String, int>> getUserStats() async {
    final userId = currentUser?.uid;
    if (userId == null) {
      return {'totalChecks': 0, 'interactions': 0, 'safe': 0};
    }

    try {
      final userData = await getUserData(userId);
      final totalChecks = userData?['totalChecks'] ?? 0;

      final snapshot = await _firestore
          .collection('interactions')
          .where('userId', isEqualTo: userId)
          .get();

      int interactions = 0;
      int safe = 0;

      for (var doc in snapshot.docs) {
        final severity = doc.data()['severity'];
        if (severity == 'High' || severity == 'Moderate') {
          interactions++;
        } else {
          safe++;
        }
      }

      return {
        'totalChecks': totalChecks,
        'interactions': interactions,
        'safe': safe,
      };
    } catch (e) {
      print('Error getting stats: $e');
      return {'totalChecks': 0, 'interactions': 0, 'safe': 0};
    }
  }
}
